File: src/ir/serialized/serialized-variable.ts
import {
  array,
  custom,
  object,
  pipe,
  readonly,
  string,
  tupleWithRest,
  type InferOutput,
} from "valibot";

import type { Span } from "../primitive/span.js";
import { referenceId$ } from "./reference-id.js";
import { scopeId$ } from "./scope-id.js";
import type { SerializedDefinition } from "./serialized-definition.js";
import { variableId$ } from "./variable-id.js";

const span$ = custom<Span>(() => true);
const serializedDefinition$ = custom<SerializedDefinition>(() => true);

// SerializedVariable carries at least one def by construction: the serializer
// filters implicit-arguments bindings (the only producer of empty defs) at
// boundary entry. tupleWithRest expresses that invariant at both runtime
// (parse rejects empty arrays) and at the type level (defs[0] narrows to
// SerializedDefinition rather than SerializedDefinition | undefined under
// noUncheckedIndexedAccess).
export const serializedVariable$ = object({
  id: variableId$,
  name: string(),
  scope: scopeId$,
  identifiers: array(span$),
  references: array(referenceId$),
  defs: pipe(
    tupleWithRest([serializedDefinition$], serializedDefinition$),
    readonly(),
  ),
});

export type SerializedVariable = InferOutput<typeof serializedVariable$>;

File: crates/unsnarl-ir/src/serialized/serialized_variable.rs
//! Serialized counterpart of `Variable`.
//!
//! `defs` is `Vec<SerializedDefinition>` rather than a "non-empty"
//! newtype: the boundary upholds the "at least one def" invariant
//! by filtering implicit-arguments bindings (the sole producer of
//! empty defs) at entry.

use serde::Serialize;

use crate::primitive::Span;
use crate::serialized::reference_id::SerializedReferenceId;
use crate::serialized::scope_id::SerializedScopeId;
use crate::serialized::serialized_definition::SerializedDefinition;
use crate::serialized::variable_id::SerializedVariableId;

#[derive(Serialize)]
pub struct SerializedVariable {
    pub id: SerializedVariableId,
    name: String,
    pub scope: SerializedScopeId,
    pub identifiers: Vec<Span>,
    pub references: Vec<SerializedReferenceId>,
    pub defs: Vec<SerializedDefinition>,
}

impl SerializedVariable {
    pub fn new(
        id: SerializedVariableId,
        name: String,
        scope: SerializedScopeId,
        identifiers: Vec<Span>,
        references: Vec<SerializedReferenceId>,
        defs: Vec<SerializedDefinition>,
    ) -> Self {
        assert!(
            !name.is_empty(),
            "SerializedVariable.name must be non-empty"
        );
        Self {
            id,
            name,
            scope,
            identifiers,
            references,
            defs,
        }
    }
}
